src/merkle/tiered_smt/values.rs

```diff
@@ -0,0 +1,587 @@
+                    Ok(pos) => {
+                        let old_value = pairs[pos].1;
+                        pairs[pos].1 = value;
+                        Some(old_value)
+                    }
+                    Err(pos) => {
+                        pairs.insert(pos, (key, value));
+                        None
+                    }
+                }
+            }
+        }
+    }
+
+    /// Removes the key-value pair with the specified key from this entry, and returns the value
+    /// of the removed pair. If the entry did not contain a key-value pair for the specified key,
+    /// None is returned.
+    ///
+    /// If the last last key-value pair was removed from the entry, the second tuple value will
+    /// be set to true.
+    pub fn remove(&mut self, key: &RpoDigest) -> (Option<Word>, bool) {
+        match self {
+            StoreEntry::Single(kv_pair) => {
+                if kv_pair.0 == *key {
+                    (Some(kv_pair.1), true)
+                } else {
+                    (None, false)
+                }
+            }
+            StoreEntry::List(kv_pairs) => {
+                match kv_pairs.binary_search_by(|kv_pair| cmp_digests(&kv_pair.0, key)) {
+                    Ok(pos) => {
+                        let kv_pair = kv_pairs.remove(pos);
+                        if kv_pairs.len() == 1 {
+                            *self = StoreEntry::Single(kv_pairs[0]);
+                        }
+                        (Some(kv_pair.1), false)
+                    }
+                    Err(_) => (None, false),
+                }
+            }
+        }
+    }
+}
+
+/// A custom iterator over key-value pairs of a [StoreEntry].
+///
+/// For a `SingleEntry` this returns only one value, but for `ListEntry`, this iterates over the
+/// entire list of key-value pairs.
+pub struct EntryIterator<'a> {
+    entry: &'a StoreEntry,
+    pos: usize,
+}
+
+impl<'a> Iterator for EntryIterator<'a> {
+    type Item = &'a (RpoDigest, Word);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.entry {
+            StoreEntry::Single(kv_pair) => {
+                if self.pos == 0 {
+                    self.pos = 1;
+                    Some(kv_pair)
+                } else {
+                    None
+                }
+            }
+            StoreEntry::List(kv_pairs) => {
+                if self.pos >= kv_pairs.len() {
+                    None
+                } else {
+                    let kv_pair = &kv_pairs[self.pos];
+                    self.pos += 1;
+                    Some(kv_pair)
+                }
+            }
+        }
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Compares two digests element-by-element using their integer representations starting with the
+/// most significant element.
+fn cmp_digests(d1: &RpoDigest, d2: &RpoDigest) -> Ordering {
```

Comment:
The sorting order on `RpoDigest` is kind of random (we sort on Montgomery representation of field elements). This is done for efficiency reasons (i.e., to avoid converting from Montgomery to conical form). In most cases this is fine as we don't actually care about the order - but in this case, the order is important and it must be based on the canonical value.

Reply:
Ah, my bad, the trait we need is `Ord`, and that is not implemented. (it probably makes sense to have it though, since it is useful for these situations, I remember we discussed this already in the past, not sure why it was not implemented).